### ISSUE TITLE
100% cpu usage when training on GPU?

### ISSUE THREAD

**Issue by fenqingr** [Microsoft/LightGBM#748]
Hi Guys,

Thank you very much for this wonderful new tool! I installed it and tested and I found that even when I used GPU for training with no data transfer between CPU and GPU during training, CPU was constantly staying at 100% usage while my GPU used only 5%~10% usage. 

My desktop environment:

cpu: i5 2500K @4G
32G DDR3 memory
gpu: EVGA GTX 1080 Ti
mainboard : AsRock gen3 extreme3
Windows 7 professional
MS VB community 2017
R 64 version 3.4.1

My questions are:
1. Why my CPU sticked at 100% usage while I training with GPU?

2. Is it normal that GPU usage only less than 10%?

3. It took supprising ~7 miniutes to finish the epsilon data when I use 63 bin but according to GPU-Performance.md it took only 165 seconds on a GTX 1080 even thought  I used a better GPU? Was that because my CPU bottleneck because it only have 4 cores?

4. I then change setting of max_bin to 15 bins and re-run the code, however, the R output still report I use 64 bins. It seems to me that the program somehow "remembered" the parameter setting of my first run and ignored my new parameter setting. I have to restart R to apply new parameters.

#below is my R code and output:

 
memory.limit=250000 
dtrain= read.table(file="F:/epsilon_normalized",sep=":",colClasses = rep("numeric",2001)  )
 
 params=list(
      max_bin = 63,objective="binary",
 num_leaves = 255,
 num_iterations = 500,
 learning_rate = 0.1,
 tree_learner = "serial",
 task = "train", 
 min_data_in_leaf = 1,
 min_sum_hessian_in_leaf = 100, 
 sparse_threshold=1.0,
 device = "cpu",
 gpu_platform_id = 0,
 gpu_device_id = 0,
 num_thread = 4 ,sparse_threshold=1
    
 )
 
 library(lightgbm)
 dtrain  <- lgb.Dataset(data =as.matrix( dtrain[ ,-1]), label = dtrain[ ,1], free_raw_data = T)
   
 s=Sys.time()
  
 model <- lgb.train(params= params ,  dtrain ,nrounds=500 , num_leaves=255)
 
 Sys.time()-s
 #Time difference of 6.828269 mins
 
OUTPUT:

[LightGBM] [Info] Number of positive: 199823, number of negative: 200177
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 128000
[LightGBM] [Info] Number of data: 400000, number of used features: 2000
[LightGBM] [Info] Using requested OpenCL platform 0 device 0
[LightGBM] [Info] Using GPU Device: GeForce GTX 1080 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 64 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 12
[LightGBM] [Info] 2000 dense feature groups (762.94 MB) transfered to GPU in 0.597143 secs. 0 sparse feature groups.
[LightGBM] [Info] Trained a tree with leaves=255 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=255 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=255 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=255 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=255 and max_depth=12
[LightGBM] [Info] Trained a tree with leaves=255 and max_depth=12
[LightGBM] [Info] Trained a tree with leaves=187 and max_depth=18
.....................
..................
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=187 and max_depth=17
 
>  Sys.time()-s
Time difference of 6.828269 mins

Any help? Thank you very much!

**Comment by guolinke** [comment]
It is normal since the GPU mode still will use cpu . you can set num_threads to reduce the cpu usage.

**Comment by guolinke** [comment]
for 2 & 3, i think @huanzhang12 can answer you.
for 4, can you provide a re-produce code ? @Laurae2 do you have any idea about it ?

**Comment by huanzhang12** [comment]
@fenqingr Yes, I believe the CPU becomes the bottleneck in your case. i7 2500k is back from 2011, and can't keep feeding the 1080 Ti with enough workload (some steps need to be done on CPU). I recommend using a recent CPU with at least 8 cores (for example, AMD Ryzen 1700) to pair with your GPU.

I am currently working on moving more workload to GPU to reduce the CPU requirement. So it might work better in a future release of LightGBM.

**Comment by fenqingr** [comment]
Thank you guys! You are awesome! I reset num_thread to 1 and now only use 25% of CPU and the run time only about 15% slower! (Does this mean actually most CPU time are actually running empty loops even if I set num_thread=4?).

For question #4, please see the code below. I first set to 16 bin and then changed to 64 bin but the output suggested #of bin remained 16 after change the bin setting.

set.seed(101)
x=matrix(rnorm(1000000),nrow=10000,ncol=100)
y=ifelse(runif(10000)>0.8,1,0)
library(lightgbm)
dtrain  <- lgb.Dataset(data =x, label =y, free_raw_data = T)

params=list(
  max_bin = 15,objective="binary",
  num_leaves = 32, 
  num_iterations = 5,
  device = "gpu"
  
)


model.use.16bin <- lgb.train(params= params ,  dtrain , gpu_use_dp=F  )

output:

[LightGBM] [Info] Number of positive: 1979, number of negative: 8021
[LightGBM] [Info] This is the GPU trainer!!
  [LightGBM] [Info] Total Bins 1600
[LightGBM] [Info] Number of data: 10000, number of used features: 100
[LightGBM] [Info] Using GPU Device: GeForce GTX 1080 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 12
[LightGBM] [Info] 100 dense feature groups (0.50 MB) transfered to GPU in 0.003825 secs. 0 sparse feature groups.
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=10
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=8
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=10
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=10
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9



params=list(
  max_bin = 63,objective="binary",
  num_leaves = 32, 
  num_iterations = 5,
  device = "gpu"
  
)


model.use.64bin <- lgb.train(params= params ,  dtrain , gpu_use_dp=F  )

[LightGBM] [Info] Number of positive: 1979, number of negative: 8021
[LightGBM] [Info] This is the GPU trainer!!
  [LightGBM] [Info] Total Bins 1600
[LightGBM] [Info] Number of data: 10000, number of used features: 100
[LightGBM] [Info] Using GPU Device: GeForce GTX 1080 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...  **<--- still report 16 bins?????????????**
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 12
[LightGBM] [Info] 100 dense feature groups (0.50 MB) transfered to GPU in 0.005077 secs. 0 sparse feature groups.
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=10
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=8
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=10
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=10
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9

**Comment by guolinke** [comment]
@fenqingr 
the max_bin is the parameter for the Dataset. So you need to define a new lgb.Dataset in your second round training.

**Comment by fenqingr** [comment]
Thank you @guolinke ! I tried to remove the Dataset and redefine one with max_bin=64 and it worked;however, when I to define it with 128 bin , the program report 256 bin. Any insight why this happened?

The code below:

params2=list(
  max_bin = 127,objective="binary",
  num_leaves = 32, 
  num_iterations = 5,
  device = "gpu"
  
)
rm(dtrain)
dtrain=lgb.Dataset(data =x, label =y, free_raw_data = T)
model.use.127bin <- lgb.train(params= params2 ,  dtrain  )


OUTPUT:
[LightGBM] [Info] Number of positive: 1979, number of negative: 8021
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 12800
[LightGBM] [Info] Number of data: 10000, number of used features: 100
[LightGBM] [Info] Using GPU Device: GeForce GTX 1080 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...  **<-- why not 128 bins?**
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 12
[LightGBM] [Info] 100 dense feature groups (0.95 MB) transfered to GPU in 0.008912 secs. 0 sparse feature groups.
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=10
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=13
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=14
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=11
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=13
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=12
[LightGBM] [Info] Trained a tree with leaves=32 and max_depth=9

One more strange thing is , note that although in params2 , I specified num_iterations = 5, the output actually contains 10 trees.

Thanks again!

**Comment by huanzhang12** [comment]
@fenqingr There are no GPU kernels for 128 bins, so the kernel for 256 bins are used. This is normal.

**Comment by fenqingr** [comment]
I see! Thanks for the help!
